### ISSUE TITLE
Intl.Segmenter is not a constructor - unable to use Element Web on Firefox ESR 115.12.0esr (64-bit)

### ISSUE THREAD

**Issue by wodny** [element-hq/element-web#27682]
### Steps to reproduce

1. Try to open the app in Firefox ESR

#### What did you expect?

I expected the app to work.

#### What happened instead?

Got an exception in console:
```
Uncaught TypeError: Intl.Segmenter is not a constructor
    node_modules bundle.js:2
    l tslib.es6.mjs:369
    tsx MediaDeviceHandler.ts:89
    l tslib.es6.mjs:369
    ts MatrixActionCreators.ts:369
    l tslib.es6.mjs:369
    ts MatrixClientBackedController.ts:30
    l tslib.es6.mjs:369
    tsx SettingLevel.ts:20
    l tslib.es6.mjs:369
    ts ConfigSettingsHandler.ts:68
    l tslib.es6.mjs:369
    tsx IntegrationManagers.ts:184
    l tslib.es6.mjs:369
    tsx DevtoolsDialog.tsx:147
    l tslib.es6.mjs:369
    ts RecorderWorklet.ts:1
    l tslib.es6.mjs:369
    c tslib.es6.mjs:369
    O tslib.es6.mjs:369
    <anonymous> tslib.es6.mjs:369
    <anonymous> tslib.es6.mjs:369
```

This API is supported [since][1] Firefox 125. I hope that dropping support for Firefox ESR is just an untested (sic!) dependency problem and not a conscious misguided decision.

[1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Segmenter

### Operating system

Debian Linux (bookworm)

### Browser information

Firefox 115.12.0esr (64-bit)

### URL for webapp

https://app.element.io/

### Application version

Can't say - it's not starting

### Homeserver

matrix.org

### Will you send logs?

No

**Comment by t3chguy** [comment]
As per https://github.com/element-hq/element-web#supported-environments we only support the last 2 major versions of each of the listed browsers, for FF this means 126 & 127.

**Comment by olberger** [comment]
Uh, it's a joke, is it ? You intend to get rid of a large user base, don't you ?

**Comment by t3chguy** [comment]
I'm just an individual contributor, I didn't write the policy. I'm merely linking you to it as it says 

> Issues only affecting unsupported environments are closed

**Comment by olberger** [comment]
> I'm just an individual contributor, I didn't write the policy. I'm merely linking you to it as it says
> 
> > Issues only affecting unsupported environments are closed

Please consider how rude this is to close other users' issues :-/

**Comment by t3chguy** [comment]
It is my job, I am tasked with triage and applying the policies and rules that are set. One of which as linked above says 

> Issues only affecting unsupported environments are closed

Therefore I have to close it

**Comment by wodny** [comment]
> As per https://github.com/element-hq/element-web#supported-environments we only support the last 2 major versions of each of the listed browsers, for FF this means 126 & 127.

The [ESR's][1] are major versions from a dedicated branch which is important for enterprise use and security-wise. By dropping support for ESR's you are dropping Debian support.

[1]: https://support.mozilla.org/en-US/kb/choosing-firefox-update-channel

**Comment by t3chguy** [comment]
The support policy never claimed support for ESR though. https://github.com/element-hq/element-web#supported-environments it says `Last 2 major versions of ..., Firefox`, not `Firefox ESR`

**Comment by wodny** [comment]
> I'm just an [individual contributor](https://honehq.com/glossary/individual-contributor/), I didn't write the policy. I'm merely linking you to it as it says
> 
> > Issues only affecting unsupported environments are closed

Please mention the policy authors here so they can rethink their position. This is to huge of a problem to discard in 5 minutes.

**Comment by t3chguy** [comment]
@davidegirardi cc as my manager

**Comment by olberger** [comment]
@t3chguy Would you mind reopening until someone knowledgeable is able to judge ?

**Comment by t3chguy** [comment]
Unfortunately not, given the policy says it must be closed. Sorry

---

Not being able to use features for an additional approx 42 weeks based on the Firefox ESR update schedule would massively slow down feature development and be a direct detriment to our funding.

**Comment by olberger** [comment]
> Unfortunately not, given the policy says it must be closed. Sorry

Being stubborn won't help in any way. Can we appeal ?

**Comment by wodny** [comment]
> The enterprise can keep using an older version of Element until they can upgrade to a browser which does not lack the necessary APIs.

Which URL should I use to use an older Element version?

**Comment by t3chguy** [comment]
> Being stubborn won't help in any way. Can we appeal ?

Its an open source project driven by a commercial entity, we are driven by a product team which takes input from management & our customers.

> Which URL should I use to use an older Element version?

One which you host yourself based on the tarballs or docker images available, I'd be surprised any enterprise would be using `*.element.io` infrastructure given it'd be an additional privacy policy in the mix and not allow for custom configuration

**Comment by olberger** [comment]
> > I'm just an [individual contributor](https://honehq.com/glossary/individual-contributor/), I didn't write the policy. I'm merely linking you to it as it says
> > > Issues only affecting unsupported environments are closed
> 
> Please mention the policy authors here so they can rethink their position. This is to huge of a problem to discard in 5 minutes.

I just filed:
https://github.com/element-hq/element-web/issues/27684 so as to be sure the policy doesn't prevent Firefox ESR to be supported.

**Comment by obfusk** [comment]
Debian stable uses Firefox ESR. This means Element Web just broke for plenty of regular users, not just "enterprise".

**Comment by wodny** [comment]
@t3chguy:
> I'm just an [individual contributor](https://honehq.com/glossary/individual-contributor/), I didn't write the policy. I'm merely linking you to it as it says

In this case it seems you are a very specific individual contributor that made an autonomous decision and switched from graphemer to `Intl.Segmenter` which is the reason of the problems. One of the users even left you a warning about it and offered an alternative in #12617:
>> Any wins against Segmenter directly?
>
> Yes for runtime perf and compatibility (if matter). That's the same reason graphemer was originally used.
>
> Intl.Segmenter is still too new (especially in Firefox) and underperforms user implementations. It comes from inefficient binding to icu4c, both in Chrome and Safari. (To be fair, they improved a lot in very recent versions)

**Comment by t3chguy** [comment]
> autonomous decision

It was discussed internally, along with confirmation that it is supported by our supported environments. ESR was not considered whatsoever at this time given it isn't listed as a supported environment.

This isn't the first ESR-related breakage for similar reason. The ESR release is said to be every ~42 weeks, the currently major of ESR is over a year old, which locks out a lot of Web features.

> One of the users even left you a warning about it and offered an alternative in https://github.com/element-hq/element-web/issues/12617:

The warning was that it is too new especially in Firefox, but again complied with our support policy by over 150%. The primary winning is smaller bundle size. The first driver of this was https://github.com/element-hq/compound-web which moved away from graphemer as graphemer was ~30% of the entire matrix-authentication-system bundle which depends on compound-web. Given that Compound now used `Intl.Segmenter` the choice to follow the same route here was basic, as we too depend on compound-web.

**Comment by wodny** [comment]
> This isn't the first ESR-related breakage for similar reason. The ESR release is said to be every ~42 weeks, the currently major of ESR is over a year old, which locks out a lot of Web features.

Firefox 128 ESR is [scheduled][1] to release tomorrow. Hopefully, it will appear in Debian soon. Couldn't the change wait a couple of weeks until the new ESR is released instead of breaking the app?

[1]: https://whattrainisitnow.com/calendar/

> The primary winning is smaller bundle size. The first driver of this was https://github.com/element-hq/compound-web which moved away from graphemer as graphemer was ~30% of the entire matrix-authentication-system bundle which depends on compound-web.

How much was the Element Web bundle/whole downloaded code size reduced?

**Comment by obfusk** [comment]
> Last 2 major versions of Chrome, Firefox, and Edge on desktop OSes
> Linux versions for desktop devices that are actively supported by the OS vendor and receive security updates

I understand (though I do not quite agree with) the reasoning here. And can see how this is being interpreted as Firefox ESR not being supported.

But as Firefox ESR is *the only officially supported version of Firefox and receiving regular security updates* on Debian stable (and regular Firefox is not available unless installed via e.g. flatpak) I think it's very much not obvious (and very much not user-friendly) that Debian stable is not supported by element web.

**Comment by t3chguy** [comment]
Mozilla themselves recommend "Rapid Release" (default) over ESR *including for Enterprise*:

![image](https://github.com/element-hq/element-web/assets/2403652/a2dc80ed-dc07-4a6d-a636-7ca3703d5dd8)

**Comment by t3chguy** [comment]
> And can see how this is being interpreted as Firefox ESR not being supported.

It is also codified as such: https://github.com/element-hq/element-web/blob/develop/babel.config.js#L7-L12

![image](https://github.com/element-hq/element-web/assets/2403652/6940d953-39d2-48ed-bfee-3eaee07666b3)

**Comment by wodny** [comment]
> > And can see how this is being interpreted as Firefox ESR not being supported.
> 
> It is also codified as such: https://github.com/element-hq/element-web/blob/develop/babel.config.js#L7-L12

On the other hand - a quote from https://browsersl.ist/ mentioned in #27684:
> Start with the default configuration
> 
> You can pick a sound set of versions with the defaults query which is a shortcut for > 0.5%, last 2 versions, Firefox ESR, not dead. It matches recent versions of popular and supported browsers worldwide and includes Firefox Extended Support Release which is updated roughly annually.
> 
> The defaults query was thoroughly designed by the Browserslist community. It helps promote best practices and avoid common pitfalls.

**Comment by obfusk** [comment]
I have been using element web on Debian with Firefox ESR for years. It has never once warned me my browser was unsupported. Nor have I heard such from other Debian users. And most websites work perfectly fine. I do not consider it unreasonable for users to expect Firefox ESR to work (which it has for years).

And now element web simply doesn't load. At all. Which also means users cannot simply switch to a different browser or client if they need the existing session for verification and decrypting existing messages.

**Comment by t3chguy** [comment]
> On the other hand - a quote from [browsersl.ist](https://browsersl.ist/) mentioned in https://github.com/element-hq/element-web/issues/27684:

That would have us support Chrome down to 109, mobile browsers, Opera, KaiOS, UC Browser, and all other things we have intentionally excluded in preference to having access to sane Web APIs shortly after they are available. We have a supports environments thing at the **top** of our README for a reason. ESR is in a similar boat, being over a year old at this point.

> It has never once warned me my browser was unsupported.

This is because we use feature detection rather than user agent parsing, as the latter is horrible and often misled by UA faking extensions. So we never look at what browser you run, only if it supports the features we need. Right now the whitescreen is a bug, it should show you an incompatibility screen - https://github.com/element-hq/element-web/pull/27685 will fix that bug

> I do not consider it unreasonable for users to expect Firefox ESR to work (which it has for years).

So if you're on any browser which hasn't had a feature update in over a year you expect it to work just because it did at some point in the past incidentally? Or is ESR somehow special in this belief

**Comment by obfusk** [comment]
I have some sympathy for only being able to support a limited subset of recent browsers. But this breakage is *unexpected for users*. And the current messaging leaves the impression that element simply does not care they now have a broken client that worked fine a few hours ago and that they cannot even export their data from.

> So if you're on any browser which hasn't had a feature update in over a year you expect it to work just because it did at some point in the past incidentally? Or is ESR somehow special in this belief

Unlike "some old version of Chrome" Firefox ESR is *actively supported by Mozilla and Debian etc.*

**Comment by mstenta** [comment]
+1 to supporting Firefox ESR. I run Debian stable as my primary OS and use Firefox ESR.

**Comment by obfusk** [comment]
> We have a supports environments thing at the top of our README for a reason.

I'm pretty sure most regular users will not have looked at the github repo, let alone seen the README. If I simply go to element.io and follow the instructions to use element web and create an account there is no mention of supported browser versions.

**Comment by obfusk** [comment]
Again, I understand not wanting to support older browsers. I get this is "policy". But users *reasonably expect* that *actively supported browsers* like Firefox ESR will work *unless they are clearly told otherwise*.

**Comment by t3chguy** [comment]
> If I simply go to element.io and follow the instructions to use element web and create an account there is no mention of supported browser versions.

You're welcome to suggest it to the team responsible for that website https://github.com/element-hq/element.io

> But users reasonably expect that actively supported browsers like Firefox ESR will work unless they are clearly told otherwise.

What about Opera? KaiOS? UC Browser? And many more. Actively maintained still, just don't meet the feature bar we need at the time we need it.

**Comment by wodny** [comment]
> So if you're on any browser which hasn't had a feature update in over a year you expect it to work just because it did at some point in the past incidentally? Or is ESR somehow special in this belief

Yes, I would expect it to work fine. Applications having just stability and security fixes for a year seem pretty normal to me. I would understand if there was a critical problem in E2E and you had to drop support. But the reason for this bug seems trivial and easy to polyfill.

Debian and Firefox (ESR) are pretty important and special. Apart from Safari Firefox is the last non-chrome browser that works pretty well. They are important part of the ecosystem. Disabling them without a warning visible for a user seems pretty extreme.

**Comment by obfusk** [comment]
> What about Opera? KaiOS? UC Browser? And many more. Actively maintained still, just don't meet the feature bar we need at the time we need it.

Are any of those the default browser on a major Linux distro, actively supported by Mozilla just like regular Firefox, and work fine until a few hours ago?

**Comment by t3chguy** [comment]
> But the reason for this bug seems trivial and easy to polyfill.

See my earlier reply

> The first driver of this was [element-hq/compound-web](https://github.com/element-hq/compound-web) which moved away from graphemer as graphemer was ~30% of the entire matrix-authentication-system bundle which depends on compound-web. Given that Compound now used Intl.Segmenter the choice to follow the same route here was basic, as we too depend on compound-web.

So element-web switching to Intl.Segmenter isn't the fault here, its Compound, which was done to help out a smaller project which was suffering from `graphemer` being so impactful on bundle size.

> and work fine until a few hours ago?

Probably

**Comment by obfusk** [comment]
There are two issues here:

1. Firefox ESR is not supported; I've pointed out why I think it should be, but not supporting it is not necessarily an unreasonable decision.
2. Firefox ESR was working just fine and many users didn't realise it wasn't supported and didn't expect it to suddenly break today and to then be told it will not be fixed.

**Comment by obfusk** [comment]
> > and work fine until a few hours ago?
> 
> Probably

I asked for a combination of 3 things (actively supported, by a major vendor (and OS), working fine). I'm not sure what saying some other browsers satisfy the last criterion is meant to prove?

**Comment by obfusk** [comment]
Look, I get the reasoning for the policy. I get supporting Firefox ESR takes more resources. What I don't get is considering it unreasonable for Debian stable users to expect element web to work for them *when they could not reasonably have expected things might suddenly break and would then not be fixed*.

**Comment by obfusk** [comment]
Had element web warned Firefox ESR users a few weeks ago that they were using a browser that didn't support a feature required by an upcoming version, giving them time to switch to a different browser or client instead of just randomly breaking for no apparent reason (from their PoV) today, I would still be unhappy, but not as disappointed (and likely to move away from element web and recommend others to do so as well) as I am now.

**Comment by t3chguy** [comment]
Debian stable users have access to other browsers & Element Desktop, they are not unable to use Element Web via supported means

> Had element web warned Firefox ESR users a few weeks ago that they were using a browser that didn't support a feature required by an upcoming version

For a feature like that to work, you'd have to know what APIs you're going to use months in advance to warn users even if they are stepping between versions rather than hitting every single one, e.g. if someone only updates Element Web once every 2 months which is quite common for certain customers. This is a slowdown to development we can't afford as a tiny team. It'd also rely on the user not having Firefox fingerprint protections enabled, as those cause the UA to be modified.

**Comment by obfusk** [comment]
> Debian stable users have access to other browsers & Element Desktop, they are not unable to use Element Web via supported means

Yes. And it's not unreasonable to tell them to use a supported means. But many *could not have reasonably known that Firefox ESR is unsupported*. And are now unexpectedly *locked out of their existing sessions which worked fine a few hours ago*.

I'm a little disappointed that Firefox ESR is not supported. But more so that I've had to drop everything to move several people to alternatives today and some of them are quite unhappy at losing access to their existing messages because they only had one session which is no longer usable.

(I will leave it at that. I don't think turning this into a long argument will help anyone.)

**Comment by t3chguy** [comment]
> But many could not have reasonably known that Firefox ESR is unsupported.

I agree, and I think suggesting to the marketing team to update the element.io website with supported browsers would be sane, hence linking to the issue repo for that.

**Comment by franzcor** [comment]
> > Debian stable users have access to other browsers & Element Desktop, they are not unable to use Element Web via supported means
> 
> Yes. And it's not unreasonable to tell them to use a supported means. But many _could not have reasonably known that Firefox ESR is unsupported_. And are now unexpectedly _locked out of their existing sessions which worked fine a few hours ago_.
> 
> I'm a little disappointed that Firefox ESR is not supported. But more so that I've had to drop everything to move several people to alternatives today and some of them are quite unhappy at losing access to their existing messages because they only had one session which is no longer usable.
> 
> (I will leave it at that. I don't think turning this into a long argument will help anyone.)

Too late, all the people previously logged with ESR will be locked out and disappointed. This was very badly handled as element web prompted me with an update and I clicked on it thinking that it would actually not lock me out.. but I was wrong.

I think that ESR users could have been notified at least weeks ago given that It's not difficult to guess that a lot of Linux users use Debian and, obviously, a lot of them use the preinstalled browser (FF ESR).
Now what? 
I absolutely don't want to use Element Desktop (electron) and I could use Librewolf for example, yes, but I'd rather keep using my old FF ESR. Would waiting a couple of weeks for the new release fix things "automagically" ?

**Comment by t3chguy** [comment]
> Would waiting a couple of weeks for the new release fix things "automagically" ?

Yes, waiting for an ESR release which is at least 10 weeks overdue. Last cut was 4th July 2023 and they claim a release every 42 weeks.

**Comment by africalimedrop** [comment]
"A sovereign and secure communications platform." that 👏 doesn't 👏 support 👏 Tor 👏 Browser
(or Mullvad)
🤡

**Comment by aligitor** [comment]
> This was very badly handled as element web prompted me with an update and I clicked on it thinking that it would actually not lock me out.. but I was wrong.

Exactly. 

I am lucky enough to have another session running on other device so I can easily move on to other matrix client. So will my community, friends and family. Having a technical issue or a debatable policy is one thing, but neglecting a group's feedback rudely is another story, it undercuts trust.

**Comment by obfusk** [comment]
> > But many could not have reasonably known that Firefox ESR is unsupported.
> 
> I agree, and I think suggesting to the marketing team to update the element.io website with supported browsers would be sane, hence linking to the issue repo for that.

Thanks. That would be great to prevent similar issues in the future. It's still unfortunate news for Firefox ESR users, but they will at least be informed and able to choose a supported alternative.

But this change comes too late for those Firefox ESR users that are currently locked out of their sessions, with no helpful error messages telling them what to do, some of which have now lost access to existing messages. If they are lucky, they will get a new ESR release soon and it will start working again, but how would they even know this?

As you've just agreed, these users could not have known they were using an unsupported browser. Closing this issue without providing these users any fix or at least a way to migrate their existing session to a supported client may be consistent with the support policy but is likely to make them very unhappy and lose trust.

Is there really nothing element is able and willing to do here to help these users out and prevent losing that trust in a service they thought they could rely on?

**Comment by o5k** [comment]
> I'm just an [individual contributor](https://honehq.com/glossary/individual-contributor/), I didn't write the policy. I'm merely linking you to it as it says

@t3chguy You are the one who [authored the commit with the breaking change](https://github.com/matrix-org/matrix-react-sdk/commit/86a95cfff7c1ca727f1d21019f5dac35351a1673). I believe it is not wise for you to be arguing in this situation; it should be brought to another member of the team to decide on whether a single contributor pushing a change that breaks major usecases and releases (Debian, Tor Browser, Mullvad Browser, Waterfox, etc, etc, etc.) is wise or good for PR.

**Comment by alfem** [comment]
Firefox ESR is used, not only by Debian users, also by enterprises and governments (like mine) that need to check software updates carefully before making a big deployment to their users.

**Comment by t3chguy** [comment]
@alfem sounds like you would be unaffected then? Just stay on the last incidentally compatible version until the overdue ESR ships?

**Comment by t3chguy** [comment]
> As you've just agreed, these users could not have known they were using an unsupported browser.

Sure, such is always the possibility regardless of how many places end up saying it

> Is there really nothing element is able and willing to do here to help these users out and prevent losing that trust in a service they thought they could rely on?

You'd first need to get https://github.com/element-hq/compound-web to stop using Intl.Segmenter otherwise it makes no difference if element-web stops using it. This would cause problem for https://github.com/matrix-org/matrix-authentication-service which asked to get rid of graphemer in the first place.

> because of a blank page.

This is crap and will be fixed by https://github.com/element-hq/element-web/pull/27685

> it should be brought to another member of the team

I already pinged my manager in this thread, not sure what more you want? You should probably also take a look at some stats https://github.com/matrix-org/matrix-react-sdk/graphs/contributors - I think I have ought to have plenty of context to provide.

**Comment by cytrinox** [comment]
Even in the age of "This site only works with IE6", web developers where able to tell the user **why** it breaks. And I'm still curios why a chat application needs the latest greatest rocket science tech. :thinking: 

May I recommend to show a top hint bar if the app is loaded in an unsupported browser? It's not the fact that a specific browser is not supported that makes user upset - but leaving the user alone with an empty white page after clicking the update button.

**Comment by t3chguy** [comment]
> May I recommend to show a top hint bar if the app is loaded in an unsupported browser?

How do you suggest using this? UA Parsing doesn't work as privacy-conscious browsers mask the User Agent, e.g. Safari on macOS 14.5 lies and says it is running on and old macOS version (10.15) where certain features won't work (`Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.5 Safari/605.1.15`)

We use feature detection but the detection won't know what features we need until we need them, so will show the incompatibility error once the feature detection notices at which point it is too late. 

> It's not the fact that a specific browser is not supported that makes user upset - but leaving the user alone with an empty white page after clicking the update button.

The white screen is a bug with the feature detection which I already made a PR to fix

**Comment by cytrinox** [comment]
> We use feature detection but the detection won't know what features we need until we need them, so will show the incompatibility error once the feature detection notices at which point it is too late.

The usual process is:
 * Release 1.2.0
 * Develop 1.3.0 with new required feature X
 * Release 1.2.1 with feature detection for X and a hint: "Your browser is outdated and has no support for features upcoming app releases needs. Please upgrade you browser, otherwise this app may became unusable."
 * Testing & QA of 1.3.0
 * Release 1.3.0

**Comment by t3chguy** [comment]
We don't have the bandwidth to maintain multiple feature branches to maintain multiple releases like this and no QA team. The team is literally four people who each have to work on different projects. Plus the above wouldn't work if an Enterprise was only updating every couple of months, unless the `1.2.1` release was done many months before `1.3.0`. Its a crap situation but it is what it is. It doesn't help that no one reported the issue with ESR until 4 days ago, many days after that change landed on develop & staging.

**Comment by t3chguy** [comment]
The fix for the feature detection not working causing a white screen has now landed and will be included in the next release cycle

**Comment by rosemash** [comment]
I was just locked out of my account because of this this. I was using the official element instance hosted at chat.mozilla.org, refreshed, and suddenly I can no longer access my session. This is the first I've heard of this incompatibility, there was no warning. I am on debian stable and not sure how to proceed.

**Comment by t3chguy** [comment]
I have raised the chat.mozilla.org issue to the EMS team, it'll likely be a case of the owners of the Mozilla instance needing to ask for a version rollback for them to have wider compatibility if they desire

**Comment by alfem** [comment]
> @alfem sounds like you would be unaffected then? Just stay on the last incidentally compatible version until the overdue ESR ships?

I am affected, of course.
How can I stay on the last version? I noticed the error after updating the webapp. AFAIK I can not undo that.

I commented this issue in mozilla-es matrix channel (using Chrome) and they are moving this internally in their webcompat group.

**Comment by t3chguy** [comment]
@alfem sounds like you didn't follow your own policy of 

> need to check software updates carefully before making a big deployment to their users.

Websites in general can only be downgraded by the people which control them, unless you can convince your browser to pull out some old cached copy but that seems unlikely

**Comment by alfem** [comment]
> @alfem sounds like you didn't follow your own policy of

Yes, I manage linux deployments for my government. But a webapp can be updated without supervision. Perhaps you are suggesting me to block Element webapp for our users, and move to a serious native app.

**Comment by t3chguy** [comment]
I was suggesting maintaining your own instance of element-web (which is purely static hosted web files so fairly trivial) that you uphold to the same policy if you expect it to remain compatible in the same manner as the rest of the system, or a native app, that's your call.

**Comment by rosemash** [comment]
Respectfully I don't think it's reasonable to expect people using a major mainline distro to simply hold back updates for element in order to remain compatible. Stable release cycles favor few feature changes, but that's not the same thing as being outdated.

**Comment by africalimedrop** [comment]
Awakening users, consider moving to a platform that takes your concerns seriously.

These fine engineers [operate from UK jurisdiction](https://web.archive.org/web/20240610063004/https://find-and-update.company-information.service.gov.uk/company/10873661), which means their sEcuRe app is awaiting [mandatory client side scanning](https://web.archive.org/web/20240703103505/https://www.legislation.gov.uk/ukpga/2023/50/section/121/enacted?view=plain). You know, the E2EE bypassing tech that the EU and US regularly block under its variously named proposals...

The above linked law is _enacted_ and awaiting accredited software to do the actual scanning of UK based services such as this one. We're already eight or nine months downstream, so expect said software to be squeezed out as soon as the UK government can plausibly rubber stamp a lowball contract without looking reckless.

All it took is one showing that this team don't take privacy seriously (not supporting Tor Browser) and a bit of digging and yeah, the rot's so much deeper than not supporting ESR.

They wanna move fast and break shit, but won't move fast enough to achieve escape velocity from a privacy hell.

Toy platform in a nanny state.

**Comment by t3chguy** [comment]
Would have taken you all of three seconds to see that we're openly against that bill 

![image](https://github.com/element-hq/element-web/assets/2403652/359b8943-907b-432b-b4ae-af0dced6cdb5)

**Comment by ara4n** [comment]
Hi folks - speaking as the guy in charge of Element; while Firefox ESR isn't technically a supported platform and we don't test it (hence this breakage), we obviously don't want to break a large set of users.  We're looking into how best to deal with this. Meanwhile, please calm down and keep this issue for tracking resolution (otherwise I'll lock it).

**Comment by t3chguy** [comment]
@ara4n https://github.com/element-hq/element-web/issues/27684 is the issue for finding a resolution

**Comment by theres-waldo** [comment]
Note: for Firefox ESR115 users on Debian-based systems who are willing to upgrade from ESR115 to the latest stable (non-ESR) release (currently Firefox 128) from Mozilla's package repository, the following steps can be used to upgrade your existing browser profile (including your existing Element session). This allows you to continue to use your existing Element session even if it was the only one:

 1. Follow the steps at https://support.mozilla.org/en-US/kb/install-firefox-linux#w_install-firefox-deb-package-for-debian-based-distributions to install Firefox 128 from Mozilla's package repository.
 2. In your Firefox ESR instance, go to `about:profiles` and take note of the name of the currently running profile. (It will be marked with "This is the profile in use and it cannot be deleted".) If you haven't created additional profiles, this will typically be `default-esr`.
 3. Close your Firefox ESR instance.
 4. Run `firefox -p`. This starts the newly installed Firefox release and opens the profile selector.
 5. Select the profile name you noted down above. Ensure "Use the selected profile without asking at startup" is checked.
 6. Click "Start Firefox".

Note that once a profile is upgraded from ESR to Release, it likely cannot be opened with ESR again.

**Comment by divVerent** [comment]
Maybe another constructive option here would be: if the web app loads but the current browser is not supported, instead of just showing an error page, the error page should provide sufficient options to export the data so it can be used in another browser.

This means the following functions should be available:

- Either:
  - Verify other devices/browsers
- Or:
  - Generate a new recovery passphrase (i.e. reset key backup)

As long as either of these things is possible even on unsupported browsers (ideally even on very old ones, like, all that Element ever worked on - this can be achieved by implementing these only using very basic functionality), this means nobody will lose their data over this, as the crypto keys in the local profile can still be migrated to another browser or device. Which is all that really matters.

In addition, it'd make sense to show the above instruction by @theres-waldo as well, or maybe a link to a wiki page detailing various ways to migrate.

**Comment by t3chguy** [comment]
> Maybe another constructive option here would be: if the web app loads but the current browser is not supported, instead of just showing an error page, the error page should provide sufficient options to export the data so it can be used in another browser.
> ...
> As long as either of these things is possible even on unsupported browsers (ideally even on very old ones, like, all that Element ever worked on - this can be achieved by implementing these only using very basic functionality), this means nobody will lose their data over this, as the crypto keys in the local profile can still be migrated to another browser or device. Which is all that really matters.

This is great in theory but in practice quite difficult. To Verify, Export, etc etc you would need to load up quite a lot of code. The init code, React, UI, React SDK, JS SDK, Rust Crypto, WASM, etc. Unless we were to replace all of those with a 2nd implementation which reimplemented all the things in lower level code which would increase development friction, brittleness and need stringent additional testing. The code would keep growing in complexity also as it would need to comprehend all database formats as db migrations occur, and given the low level nature it'd end up being a rather large project all on its own. Not at all impossible but infeasible for a team of 3.5 people with ongoing project obligations already prescheduled for many months ahead.

**Comment by divVerent** [comment]
Then maybe we need a clear warning when using Element about the browser support policy, and that if the user cannot abide by it by using officially supported browsers, they WILL lose all their data and verifications, and thus should REALLY add a second device or make sure they REALLY never lose the recovery key.

Like, I'm a software engineer, I know to not ever rely on data stored in web browsers. But most end users do not, and given this policy of Element - which is IMHO very surprising from a user perspective, I personally thought we've long moved past "Best viewed with Internet Explorer 4 on 800x600 on an Pentium MMX with a Geforce 2 MX" - this deserves a strong warning to end users.

**Comment by obfusk** [comment]
Given that the *immediate* issue for most users here is that Firefox ESR doesn't support `Intl.Segmenter`, would it perhaps be possible to temporarily add a polyfill for that so it starts working again? Perhaps also show some kind of warning when the polyfill is needed. That doesn't mean you have committed to support Firefox ESR indefinitely, but it would unbreak things for Firefox ESR users and allow much less painful migration to a supported browser/client.

Then you also have more time to figure out whether Firefox ESR can be supported and make sure it's properly communicated to users whether their browser is supported or not so users can switch to a different browser if they need to.

**Comment by superkuh** [comment]
I will no longer use matrix. Not supporting stable browsers is wrong.

**Comment by superkuh** [comment]
I'd like to add that my saved 12 group security key I had saved in a text file doesn't work, "Could not enable key backup: Bad MAC.", so until I can get Firefox ESR running element.io client again I can't even access my account with other clients. Thanks for the zero warning it'd just stop working. At this point I might as well give up on matrix protocol entirely.

**Comment by alphapapa** [comment]
> I'd like to add that my 12 group security key I had saved in a text file doesn't work, "Could not enable key backup: Bad MAC.", so until I can get my older Firefox install running element.io client again I can't even access my account history with other clients like Hydrogen Chat. Thanks for the zero warning it'd just stop working. At this point I might as well give up on matrix protocol entirely.

FWIW, my saved key works when pasted into a fresh browser instance.

**Comment by hartnegg** [comment]
If you _knowingly_ break compatibility with the latest ESR version, even without warning anybody, you should at least change your webpage "Supported Environments". It does not, but should specificly say, that you support only the Rapid Release versions.

**Comment by knarrff** [comment]
It seems that mostly this comes from someone, intentionally or not, not specifying ESR versions as supported and, probably someone else, using that statement to argue for a change that locks out large parts of the user base. To me, this suggests that original problem is the intended list of supported browser versions: who can make decisions about that and how can we reach/influence those people - making them aware that this is ruinous to their project?

**Comment by divVerent** [comment]
Also, I think we're conflating two issues here:

- The Element client's code base: that can of course have any kind of browser requirements it wants, and it's up to admins to update or not, etc. - here it'd be nice to simply declare new API requirements so admins can make the decision in an informed manner.

- The app.element.io web client - this should be administrated in a way to minimize disruption. E.g. before a browser-breaking change is rolled out, users should be notified some weeks in advance so they are given time to export their data or upgrade their browser.

Then there's also the third issue that the browser requirements at https://github.com/element-hq/element-web?tab=readme-ov-file#supported-environments  aren't clearly enough written - most Firefox ESR users probably think they ARE running one of "Last 2 major versions of Chrome, Firefox, and Edge on desktop OSes".

So maybe this should be worded clearer to explicitly exclude ESR?

But also, I've never seen these requirements ever before this thread. Signing up at app.element.io doesn't show them. Maybe it should?

**Comment by knarrff** [comment]
> So maybe this should be worded clearer to explicitly exclude ESR?

I think the point a lot of the commenters here tried to bring across is that it ought to explicitly include, not exclude them.

**Comment by melroy89** [comment]
Unable to use Element Web with Floorp, because you dropped support of ESR. This is insane.

**Comment by langleyd** [comment]
Thanks again to all for the feedback and thoughts on this topic.

Until now Firefox ESR was outside of our support policy, and was therefore not a browser we tested for. This coupled with a recent breakage of the “Unsupported Browsers” screen, meant users were left in a broken state. Apologies to those users left in this state during the last release while a decision was made on the possibility of adding support.

Element has decided to add support on a _best effort_* basis for the latest Firefox ESR and Chrome/Edge Extended Stable browsers, the timeframe for which will be extended until the new version of Firefox ESR lands in Debian stable plus one additional release cycle(2 weeks). We have fixed the Unsupported Browsers screen and also and have also added a [code change](https://github.com/element-hq/element-web/pull/27803) for the upcoming Release Candidate that will allow users on Firefox ESR 115 to continue to the product from the Unsupported Browsers screen. We will also be improving the in-app experience to better inform users ahead of time that support for their browser will cease.

\* What does “best-effort” mean in practice?
- The wider Element Products(including Element Call and the Enterprise Server Suite) do still not officially support these browsers. Element Call currently works with Firefox ESR 128 but a break in the future may be unavoidable (E.g. If Element called needed to use a modern api to support End-to-End encrypted VoIP).
- The element web project and its contributors should keep the client functioning and gracefully degrade where other sibling features like Element Call may not function.
- We will invest in tooling and processes to try ensure the extended support browsers continue to function for the period outlined above though this may be challenging; for example our primary tool for such testing, Playwright, does not currently support older Firefox releases (see[ https://github.com/microsoft/playwright/issues/31163](https://github.com/microsoft/playwright/issues/31163)).
